src/NHibernate.Test/Criteria/Lambda/ExpressionProcessorFindValueFixture.cs

```diff
@@ -0,0 +1,321 @@
+using System;
+using System.Linq.Expressions;
+using System.Reflection;
+using NHibernate.Impl;
+using NUnit.Framework;
+using Expression = System.Linq.Expressions.Expression;
+
+namespace NHibernate.Test.Criteria.Lambda
+{
+	public static class DateTimeExtensions
+	{
+		public static long ToLongDateTime(this DateTime date)
+		{
+			return Convert.ToInt64(date.ToString("yyyyMMddHHmmss"));
+		}
+	}
+
+	[TestFixture]
+	public class ExpressionProcessorFindValueFixture
+	{
+		private static object GetValue<T>(Expression<Func<T>> expression)
+		{
+			try
+			{
+				return ExpressionProcessor.FindValue(expression.Body);
+			}
+			catch (TargetInvocationException e)
+			{
+				throw e.InnerException;
+			}
+		}
+
+		private static int GetIntegerDate()
+		{
+			return Convert.ToInt32(DateTime.Now.ToString("yyyyMMdd"));
+		}
+
+		[Test]
+		public void StaticPropertyInstanceMethodCall()
+		{
+			var actual = GetValue(() => DateTime.Now.ToString("yyyyMMdd"));
+			var expected = DateTime.Now.ToString("yyyyMMdd");
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void StaticPropertyInstanceMultipleMethodCall()
+		{
+			var actual = GetValue(() => DateTime.Now.AddDays(365).ToString("yyyyMMdd"));
+			var expected = DateTime.Now.AddDays(365).ToString("yyyyMMdd");
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void StaticPropertyInstanceMethodCallThenCast()
+		{
+			var actual = GetValue(() => Convert.ToInt32(DateTime.Now.AddDays(365).ToString("yyyyMMdd")));
+			var expected = Convert.ToInt32(DateTime.Now.AddDays(365).ToString("yyyyMMdd"));
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void StaticMethodCall()
+		{
+			var actual = GetValue(() => GetIntegerDate());
+			var expected = GetIntegerDate();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void ExtensionMethodCall()
+		{
+			var date = DateTime.Now;
+			var actual = GetValue(() => date.ToLongDateTime());
+			var expected = date.ToLongDateTime();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void NestedPropertyAccess()
+		{
+			var animal = new { Snake = new { Animal = new { Name = "Scorpion" } } };
+			var actual = GetValue(() => animal.Snake.Animal.Name);
+			var expected = animal.Snake.Animal.Name;
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void GuidToStringCast()
+		{
+			var guid = Guid.NewGuid();
+			Expression<Func<string>> expression = () => $"{guid}";
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var lambdaExpression = Expression.Lambda(expression).Compile();
+			var expected = ((dynamic) lambdaExpression.DynamicInvoke()).Invoke();
```

Comment:
@gokhanabatay, @bahusoid why do we here double-compile lambda? Why not just `expression.Compile().Invoke()`?

Reply:
@hazzik  it's not necessary my mistake I have fixed